language/move-lang/src/parser/syntax.rs

```diff
@@ -257,65 +282,80 @@ fn parse_module_name(tokens: &mut Lexer) -> Result<ModuleName, Error> {
-    }
-
-    consume_token(tokens, Tok::ColonColon)?;
-    let name = parse_module_name(tokens)?.0;
+    consume_access_colon_colon(tokens, start_loc, "after an address in a module identifier")?;
+    let module = parse_module_name(tokens)?;
     let end_loc = tokens.previous_end_loc();
     let loc = make_loc(tokens.file_name(), start_loc, end_loc);
-    Ok(ModuleIdent {
-        locs: (loc, name.loc),
-        value: (address, name.value),
-    })
+    Ok(sp(loc, ModuleIdent_ { address, module }))
 }
 
 // Parse a module access (a variable, struct type, or function):
 //      ModuleAccess =
 //          <Identifier>
-//          | <ModuleName> "::" <Identifier>
-//          | <ModuleIdent> "::" <Identifier>
-fn parse_module_access<F: FnOnce() -> String>(
+//          | <LeadingNameAccess> "::" <Identifier>
+//          | <LeadingNameAccess> "::" <Identifier> "::" <Identifier>
+fn parse_name_access_chain<'a, F: FnOnce() -> &'a str>(
     tokens: &mut Lexer,
     item_description: F,
-) -> Result<ModuleAccess, Error> {
+) -> Result<NameAccessChain, Error> {
     let start_loc = tokens.start_loc();
-    let acc = match tokens.peek() {
-        Tok::IdentifierValue => {
-            // Check if this is a ModuleName followed by "::".
-            let m = parse_identifier(tokens)?;
-            if match_token(tokens, Tok::ColonColon)? {
-                let n = parse_identifier(tokens)?;
-                ModuleAccess_::ModuleAccess(ModuleName(m), n)
-            } else {
-                ModuleAccess_::Name(m)
-            }
-        }
+    let access = parse_name_access_chain_(tokens, item_description)?;
+    let end_loc = tokens.previous_end_loc();
+    Ok(spanned(tokens.file_name(), start_loc, end_loc, access))
+}
 
-        Tok::NumValue => {
-            let m = parse_module_ident(tokens)?;
-            consume_token(tokens, Tok::ColonColon)?;
-            let n = parse_identifier(tokens)?;
-            ModuleAccess_::QualifiedModuleAccess(m, n)
+// Helper for parse_name_access_chain
+fn parse_name_access_chain_<'a, F: FnOnce() -> &'a str>(
+    tokens: &mut Lexer,
+    item_description: F,
+) -> Result<NameAccessChain_, Error> {
+    let start_loc = tokens.start_loc();
+    let ln = parse_leading_name_access_with_description(tokens, item_description)?;
+    let ln = match ln {
+        // A name by itself is a valid access chain
+        sp!(_, LeadingNameAccess_::Name(n1)) if tokens.peek() != Tok::ColonColon => {
+            return Ok(NameAccessChain_::One(n1))
         }
+        ln => ln,
+    };
 
+    consume_access_colon_colon(
+        tokens,
+        start_loc,
+        "after an address in a module access chain",
+    )?;
+    let n2 = parse_identifier(tokens)?;
+    if tokens.peek() != Tok::ColonColon {
+        return Ok(NameAccessChain_::Two(ln, n2));
+    }
+
+    consume_token(tokens, Tok::ColonColon)?;
+    let n3 = parse_identifier(tokens)?;
+    Ok(NameAccessChain_::Three(ln, n2, n3))
+}
+
+fn consume_access_colon_colon(
```

Comment:
Can we generalize this? It's very close to `consume_token` but with the addition of the `case` argument. Maybe `consume_token` and `unexpected_token_error` should have an optional `case` argument? Or at least we could add this as a variant `consume_token_with_case`. I suggest placing this function next to `consume_token` as well.